### ISSUE TITLE
Feature Request: populateType (joinType)

### ISSUE THREAD

**Issue by RWOverdijk** [balderdashy/waterline#597]
I vote for adding the possibility to choose a `populateType` on `.populate()` calls.

When supplying criteria for a populate, I sometimes wish to omit the entire result if none match _(regular join)_. Currently, it looks like waterline, by default, uses a `left outer join` approach. This is fine for default behavior, but sometimes it might be useful to be able to filter results. Take this example:

* Model `User` has basic properties. A `firstName`, `lastName`, `username` and `dateOfBirth`.
* Model `UserProfile` has a oneToOne association with `User`, and has properties `length`, `eyeColor` and `hairColor`.
* Model `UserMedia` has a ManyToOne association with `User`, too. It holds a `type` and `file` property.

Now, I want to look for:

```javascript
var userModel = sails.models.user
  , findQuery = userModel.find({firstName: {startsWith: 'a'}});

// Populate UserProfile
findQuery.populate('userprofile', {eyeColor: 'green'});

// Populate UserMedia
findQuery.populate('usermedia', {type: 'avatar'});

// Get the results
findQuery.exec(function(error, result) {
  // I'll have too many results here.
});
```

I'll get all usernames starting with `a`, including those that do not have green eyes, and / or do not have an avatar. Here, I'd like to be able to say something like `populateType: 'strict'` (or SQL terminology, outer) and exclude all rows entirely if they don't have a match.

**Comment by RWOverdijk** [comment]
Has anyone solved join criteria like this yet (without writing a custom query)?

**Comment by Globegitter** [comment]
@RWOverdijk yep that is something I would ideally have as well in my all. I think the proposed syntax would work well and was what I tried initially using 'not' : null. 

Are you still open to submit a PR yourself? Are you just waiting for a couple of pointers before getting started?

**Comment by RWOverdijk** [comment]
@Globegitter I'm waiting for pointers to be honest. My proposed syntax works, if I change the "where" builder. It would then build the where criteria for the table name as proposed. However, I think it'd be weird you're not allowed to set them in the criteria of `.populate()`. So I'm basically hoping for some good solid feedback with pointers. Even if we only build support for sequel based adapters.

**Comment by Globegitter** [comment]
I think @particlebanana was clear that he did not want to add more to the `.populate()` syntax and I personally think from an API standpoint it could fit in either, but not in both. And looking at the syntax I think what is currently proposed is cleaner. You have all the filter functionality in the `where` and keep the `.populate` clean and simple.

I am not an expert on Mongo, but since it does support the dot syntax to filter associations I think it would belong further up in waterline. As @tjwebb already suggested.

I think the first spot to look at is https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/finders/basic.js#L182, possibly here https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/finders/basic.js#L299
But I would suggest for you to get a feel for the code, either start writing some tests that describe the behaviour you want and then, when running the tests plaster the function with `console.logs` at strategic points printing the criterias, values, return results etc.
Otherwise you could do the same, just within one of your local sails apps and just modify the code within `node_modules/sails/node_modules/waterline`.

Or to do it even cleaner you could use http://nodejs.org/api/debugger.html or https://github.com/node-inspector/node-inspector but dont' have much experience with either, personally.

**Comment by atiertant** [comment]
https://github.com/balderdashy/waterline/issues/988 => could be done by defining multiple relations with different criteria

**Comment by RWOverdijk** [comment]
I've built association where clause support, so in a workaround I could say assoc: {id : { not : null}}. This issue is less important to me personally now.

**Comment by sailsbot** [comment]
Thanks for posting, @RWOverdijk.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
